### PR TITLE
[WIP] Add support to validate using JSON schema

### PR DESCRIPTION
**Pull request by cjh1**

### CODE REVIEW

**Comment by cjh1** [comment]
@mgrauer PTAL, this is WIP, this need to work out how to support $ref resolution ...

File: girder/models/model_base.py [hunk 1]
```diff
@@ -192,8 +206,13 @@ def validate(self, doc):
         :param doc: The document to validate before saving to the collection.
         :type doc: dict
         """
-        raise Exception('Must override validate() in %s model.'
-                        % self.__class__.__name__)  # pragma: no cover
+        if self.schema:
+            try:
+                jsonschema.validate(doc,  self.schema)
```
Comment (mgrauer) [inline girder/models/model_base.py]: two spaces here

File: requirements.txt [hunk 2]
```diff
@@ -11,6 +11,7 @@ python-dateutil==2.5.3
 pytz==2016.4
 requests==2.10.0
 six==1.10.0
+jsonschema==2.5.1
```
Comment (mgrauer) [inline requirements.txt]: I think you would need to add this (modified in some way to make it less specific) to the setup.py also

Comment (cjh1) [inline requirements.txt]: Good catch.

File: girder/models/model_base.py [hunk 3]
```diff
@@ -61,8 +63,20 @@ def __init__(self):
...
+    def _validateSchema(self):
+        """
+        Validate schema if one has been provided.
+        """
+        if self.schema:
+            try:
+                validator = jsonschema.validators.validator_for(self.schema)
+                validator.check_schema(self.schema)
+            except jsonschema.exceptions.SchemaError as se:
+                raise Exception('Invalid JSON schema: %s' % se.message)
```
Comment (mgrauer) [inline girder/models/model_base.py]: Maybe you could add a bit of debugging here, like the name of the model at least.

I've had a hard time hunting down these json schema validation errors in the past.

Comment (cjh1) [inline girder/models/model_base.py]: Sure, good idea

**Comment by mgrauer** [comment]
@cjh1 I like your approach here for adding schemas to models and validating them.

I'm having slight concerns about the whole idea and wanted to at least describe them.  We are storing models in Mongo, which is schemaless, and this will start introducing schemas to models, which could mean issues like data migration (e.g., update saved models when a schema changes).  We would also want to ensure that developers add schemas that don't restrict the ability to add new fields, which is a feature Girder relies upon to allow plugins to extend core models.

Of course, this is written as an opt-in, so you aren't forcing anyone to adopt schemas for their models.

File: girder/models/model_base.py [hunk 4]
```diff
@@ -192,8 +206,13 @@ def validate(self, doc):
...
         """
-        raise Exception('Must override validate() in %s model.'
-                        % self.__class__.__name__)  # pragma: no cover
+        if self.schema:
+            try:
+                jsonschema.validate(doc, self.schema)
+            except jsonschema.ValidationError as ve:
+                raise ValidationException(ve.message)
+
+        return doc
```
Comment (cjh1) [inline girder/models/model_base.py]: One issue here is this removes the requirement to override the validate method ...

Comment (zachmullen) [inline girder/models/model_base.py]: That's no problem.

**Comment by cjh1** [comment]
@mgrauer Not sure how useful the latest addition is, but was experimenting with validation of incoming requests.

**Comment by cjh1** [comment]
Another possibility is to add another decorator to do validation of parameters on an endpoint. Something like:

```
@validateJson(schema=..., param='name')
```

Any value? Thoughts?

**Comment by zachmullen** [comment]
That would be great for handling requests that accept JSON in the body of the request.

**Comment by cjh1** [comment]
For the request body it would be nice to have a decorator, that injects the result of calling getBodyJson(...), something like:

```
@jsonBody(schema=...)
def myEndpoint(body, ...)
```

**Comment by zachmullen** [comment]
That is cool, though I fear we may be starting to have too many decorators on our REST endpoints. Let's hold off on that for now and evaluate later on whether we want to make it into a decorator.

**Comment by zachmullen** [comment]
@cjh1 PTAL at #1701 

I think this feature would best be implemented within the newly proposed `autoDescribeRoute`; there is a new method on the description called `jsonParam`, and I envision (but have not yet implemented) a `jsonSchema` kwarg to that which could validate both body JSON and JSON in form or query params. What do you think?

**Comment by cjh1** [comment]
@zachmullen Yes, I saw #1701 and though of this branch. I think when I started the schema work we discussed using a decorator, but there where concerns about there proliferation, as we are adding this one I think its a natural fit for this functionality. I would be happy to add this on a follow up PR to #1701.

**Comment by zachmullen** [comment]
We now have json schema support in autoDescribeRoute, so I'm closing this. If further action is needed, go ahead and open a PR against latest master.

**Pull request closed without merging**